---
title: "MBioMed_Practical #1: Gene Expression Analysis"
author: "Christoforos Nikolaou"
date: "November 9, 2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Practical #1: RNA Expression Analysis
## 1.1 Data Normalization
All measurements were converted to gene space by calculating mean probeset values referring to the same genes. Only values from genes that were measured in all samples were finally included in the dataset which consisted of 18704 common measured genes in all 66 samples. 
(Supplementary File: Exp1234_raw_common18704genes_antiTNF.tsv)

## 1.2 ANOVA with Dunnett's test
The code below describes the application of ANOVA and Dunnett's test on the raw (de-logged) microarray data. An initial step of quantile normalization is followed by the ANOVA. As the execution can take a while the user can skip this section as its output is provided as supplementary file. If the user wants to execute the following chunk, the option "include" should be changed to TRUE.
```{r Analysis of expression data with ANOVA coupled with Dunnett's test, include=FALSE}
library(multcomp)
data<-read.delim("Raw_common18704genes_antiTNF.tsv", header=T, sep="\t")
boxplot(data[,2:67])
genes=data$Gene
library("preprocessCore")
normalize.quantiles(as.matrix(data[,c(2:67)]),copy=TRUE)->matrix
boxplot(matrix)
# A should be control!!!! # Order should be alphabetical
# Part 1: Using Wt as control
Group <- factor(c(rep("A_Wt",10), rep("B_Tg",13), rep("C_RemProph",3), rep("D_RemTher",10), rep("E_HumTher",10), rep("F_EnbTher",10), rep("G_CimTher",10)))
x<-vector(length(matrix[,1]),mode="numeric")
lfcTgWt=vector(length(matrix[,1]),mode="numeric")
pTgWt=vector(length(matrix[,1]),mode="numeric")
lfcRemPWt=vector(length(matrix[,1]),mode="numeric")
pRemPWt=vector(length(matrix[,1]),mode="numeric")
lfcRemWt=vector(length(matrix[,1]),mode="numeric")
pRemWt=vector(length(matrix[,1]),mode="numeric")
lfcHumWt=vector(length(matrix[,1]),mode="numeric")
pHumWt=vector(length(matrix[,1]),mode="numeric")
lfcEnbWt=vector(length(matrix[,1]),mode="numeric")
pEnbWt=vector(length(matrix[,1]),mode="numeric")
lfcCimWt=vector(length(matrix[,1]),mode="numeric")
pCimWt=vector(length(matrix[,1]),mode="numeric")
for(i in 1:length(matrix[,1])){
	if (i %% 100 == 0) {cat("iteration =", i, "\n")}
	Value <- unname(unlist(matrix[i,1:66])) 
#
	data <- data.frame(Group, Value)
	aov <- aov(Value ~ Group, data)
	set.seed(123456789)
	Dunnett<-summary(glht(aov, linfct=mcp(Group="Dunnett")))
#
	TgWt=Dunnett[[10]]$coefficients[1]
		if(2^TgWt>1) {lfcTgWt[i]=log2(2^TgWt)}
		if(2^TgWt<1) {lfcTgWt[i]=-log2(1/2^TgWt)}
	pTgWt[i]=Dunnett[[10]]$pvalues[1]
#
	RemPWt=Dunnett[[10]]$coefficients[2]
		if(2^RemPWt>1) {lfcRemPWt[i]=log2(2^RemPWt)}
		if(2^RemPWt<1) {lfcRemPWt[i]=-log2(1/2^RemPWt)}
	pRemPWt[i]=Dunnett[[10]]$pvalues[2]
#
	RemWt=Dunnett[[10]]$coefficients[3]
		if(2^RemWt>1) {lfcRemWt[i]=log2(2^RemWt)}
		if(2^RemWt<1) {lfcRemWt[i]=-log2(1/2^RemWt)}
	pRemWt[i]=Dunnett[[10]]$pvalues[3]
#
	HumWt=Dunnett[[10]]$coefficients[4]
		if(2^HumWt>1) {lfcHumWt[i]=log2(2^HumWt)}
		if(2^HumWt<1) {lfcHumWt[i]=-log2(1/2^HumWt)}
	pHumWt[i]=Dunnett[[10]]$pvalues[4]
#
	EnbWt=Dunnett[[10]]$coefficients[5]
		if(2^EnbWt>1) {lfcEnbWt[i]=log2(2^EnbWt)}
		if(2^EnbWt<1) {lfcEnbWt[i]=-log2(1/2^EnbWt)}
	pEnbWt[i]=Dunnett[[10]]$pvalues[5]
#
	CimWt=Dunnett[[10]]$coefficients[6]
		if(2^CimWt>1) {lfcCimWt[i]=log2(2^CimWt)}
		if(2^CimWt<1) {lfcCimWt[i]=-log2(1/2^CimWt)}
	pCimWt[i]=Dunnett[[10]]$pvalues[6]
}
# Joing data in one data.frame
datadunnett<-data.frame(Gene=genes, "log2FC_Tg/Wt"=lfcTgWt, "pvalue_Tg/Wt"=pTgWt, "log2FC_RemP/Wt"=lfcRemPWt, "pvalue_RemP/Wt"=pRemWt, "log2FC_Rem/Wt"=lfcRemWt, "pvalue_Rem/Wt"=pRemWt, "log2FC_Hum/Wt"=lfcHumWt, "pvalue_Hum/Wt"=pHumWt, "log2FC_Enb/Wt"=lfcEnbWt, "pvalue_Enb/Wt"=pEnbWt, "log2FC_Cim/Wt"=lfcCimWt, "pvalue_Cim/Wt"=pCimWt) 

#Exporting data.frame to a new file called:Dunnett_antiTNF_vs_Wt_and_Tg.tsv
write.table(datadunnett, file="Dunnett_antiTNF_vs_Wt_and_Tg.tsv", quote=F, row.names=F, sep="\t")
```

## 1.3 Differential Expression Analysis
Reading of log-foldchange data after an ANOVA followed by Dunnett's test. A function for selection of DE genes according to logFC and p-value thresholds is introduced and applied on each condition. The dataset is exported from the previous code chunk, but is also provided directly as Supplementary File: Dunnett_antiTNF_vs_Wt_and_Tg.tsv.
```{r DEG calculations}
data<-read.delim("Dunnett_antiTNF_vs_Wt_and_Tg.tsv")[,1:13]
cim<-62:67
enb<-52:57
hum<-42:47
rem<-28:34
remp<-25:27
tg<-12:18
wt<-2:8
corindexes<-c(wt, tg, remp, rem, hum, enb, cim)
#
# Function to get Differentially Expressed Genes with given thresholds for abs(log(FC)) and p-value
getdeg<-function(data,logfc,pvalue){which(abs(data[,2])>=logfc & data[,3]<=pvalue)->i; return(i)}
tgdegs<-0
tgdegs<-c(tgdegs,getdeg(data[,1:3], 1, 0.05))
tgdegs<-tgdegs[2:length(tgdegs)]
tgdegnames<-data[tgdegs,1]
#
tgdegs<-0;
for(i in 1:6){
  x<-i*2
  y<-i*2+1
tgdegs<-c(tgdegs,getdeg(data[,c(1,x,y)], 1, 0.05))
}
tgdegs<-unique(tgdegs)
tgdegs<-tgdegs[2:length(tgdegs)]
degnames<-data[tgdegs,1]
#
degnames
```

## 2.4 Volcano Plots
```{r, echo=T, eval=T}
datadunnett<-read.delim("Dunnett_antiTNF_vs_Wt_and_Tg.tsv")[,1:13]
volcanoPlot<-function(data, lfcthres, pvalthres){
which(data[,2]>=lfcthres & data[,3]<=pvalthres)->up
which(data[,2]<=-lfcthres & data[,3]<=pvalthres)->down
which(abs(data[,2])<lfcthres | data[,3]>pvalthres)->nochange 
plot(data[nochange,2], -log10(data[nochange,3]), type="p", pch=19, col="grey", xlab="log2(FC)", ylab="-log10(p-value)", main="Deregulation vs Control", xlim=c(min(data[,2]),max(data[,2])), ylim=c(0, 15));
lines(data[up,2], -log10(data[up,3]), type="p", pch=19, col="dark red");
lines(data[down,2], -log10(data[down,3]), type="p", pch=19, col="dark blue");
legend("topleft", c("no change", "Up", "Down"), fill=c("grey","dark red", "dark blue"), bty="n")
}
volcanoPlot(datadunnett[,1:3], 1, 0.05)
```

## 1.6 Clustering of differentially expressed genes
Agglomerative hierarchical clustering using Ward's minimum variance criterion for both genes and conditions.
```{r TgDeg Heatmap}
library(gplots)
data<-read.delim("Dunnett_antiTNF_vs_Wt_and_Tg.tsv")[,1:13]
hcols<-colorRampPalette(c("steelblue4","white","firebrick4"))(256)
pcols<-c("firebrick4", "red", "dark orange", "gold","dark green", "dodgerblue", "blue", "magenta", "darkorchid4")
m<-heatmap.2(as.matrix(data[tgdegs,seq(from=2, to=12, by=2)]), col=hcols, labRow="", tracecol="black", vline=0, mar=c(8,5), hclustfun = function(x) hclust(x,method = 'ward.D2'))
mt<-as.hclust(m$rowDendrogram);cutree(mt, k=8)->tgcluster;table(tgcluster)
treatments<-c("Disease", "Infliximab3w", "Infliximab6w", "Adalimumab6w", "Etanercept6w", "Certolizumab6w")
heatmap.2(as.matrix(data[tgdegs,seq(from=2, to=12, by=2)]), col=hcols, labRow="", tracecol="black", vline=0, mar=c(10,2), hclustfun = function(x) hclust(x,method = 'ward.D2'), density.info="none", key.xlab = "log2(FC)", labCol = treatments, RowSideColors = pcols[tgcluster])
```

## 1.5 Beanplots for DE gene clusters
```{r TgDeg Beanplots}
library(beanplot)
par(mfrow=c(2,4))
par(mar=c(3,5,3,1))
for (i in c(6,5,4,3,2,8,7,1)){
  j<-as.numeric(names(which(tgcluster==i)))
  beanplot(data[j,c(2,4,6,8,10,12)], names=c("Dis","Infl3","Infl6","Adal6","Etan6", "Cert6"), what=c(0,1,1,0), col=pcols[i], horizontal=T, main=paste("Cluster", i), las=1, ylim=c(-3,3));abline(v=0, lty=3, lwd=1, col="black")
}
```


## 1.6 Functional Analysis
Functional Analysis was performed at the levels of GO terms (BP: Biological Process, MF: Molecular Function, CC: Cellular Component), KEGG pathways and Transcription Factor targets. The gProfileR library was used in all cases.
The code below is used to re-create Figure 1B but the full list of enriched terms can also be compiled by storing the output in stable files instead of temporary (temp).
```{r, Disease-Associated Gened Functional Analysis}
library(gProfileR)
totalresults<-0
totalcols<-0
funcanal<-function(genes, clusters, species){
  funclist<-list()
  cl<-as.numeric(names(table(cluster)))
  for (i in 1:length(cl)){
    querygenes<-as.character(genes[which(cluster==cl[i])])
    funcs<-gprofiler(query=querygenes, organism = species, significant=T)[,c(9,12,4,3)]
    funclist<-list(funclist, c(paste("Cluster",cl[i]), funcs))
  }
  return(funclist)
}
cl<-as.numeric(names(table(tgcluster)))
#
for(i in c(6,5,4,3,2,7,1)){
gprofiler(query=as.character(tgdegnames[which(tgcluster==cl[i])]), organism="mmusculus", significant=T, src_filter="GO:BP")[,c(9,12, 4,3)]->gobp
gprofiler(query=as.character(tgdegnames[which(tgcluster==cl[i])]), organism="mmusculus", significant=T, src_filter="GO:MF")[,c(9,12, 4,3)]->gomf
gprofiler(query=as.character(tgdegnames[which(tgcluster==cl[i])]), organism="mmusculus", significant=T, src_filter="GO:CC")[,c(9,12, 4,3)]->gocc
gprofiler(query=as.character(tgdegnames[which(tgcluster==cl[i])]), organism="mmusculus", significant=T, src_filter="TF")[,c(9,12, 4,3)]->tf
gprofiler(query=as.character(tgdegnames[which(tgcluster==cl[i])]), organism="mmusculus", significant=T, src_filter="KEGG")[,c(9,12, 4,3)]->kegg
rbind(kegg, tf, gobp, gomf, gocc)->all
# write to file temp (temporary)
write.table(all, file="temp", row.names = F, sep="\t", quote=F, col.names = T)
# read from temp
results<-read.delim("temp", header=T, sep="\t")
grep("TF:", results$term.id)->tf
grep("GO:", results$term.id)->go
grep("KEGG:", results$term.id)->kegg
results[kegg,]->kegg
results[tf,]->tf
results[go,]->go
kegg[order(kegg$p.value),]->kegg
go[order(go$p.value),]->go
tf[order(tf$p.value),]->tf
strsplit(as.character(tf$term.name), ": ")->ll
sapply(ll, "[[", 2)->ll
strsplit(as.character(ll), ";")->ll
sapply(ll, "[[", 1)->tf$term.name
if(length(tf$term.id)>0){
unique(tf$term.name)->uniqtf
tfout<-0;for(ik in 1:length(uniqtf)){which(as.character(tf$term.name)==as.character(uniqtf[ik]))->nn; tfn<-tf[nn,]; which(tfn$p.value==min(tfn$p.value))->inn; tfout<-rbind(tfout,head(tfn[inn,],1))};tfout[2:length(tfout[,1]),]->tf
}
results<-rbind(head(kegg,10),head(go,10),head(tf,10))
totalresults<-rbind(totalresults,results)
n<-length(results$term.id)
totalcols<-c(totalcols, rep(pcols[i],n))
}
totalresults<-totalresults[2:length(totalresults[,1]),]
totalcols<-totalcols[2:length(totalcols)]
par(mar=c(5,15,1,2));barplot(rev(-log10(totalresults$p.value[75:126])),xlab="-log10(p-value)",ylab="",cex.main=1.3, cex.lab=0.9, cex.axis=0.9, main="Under-Expressed Clusters", col=rev(totalcols[75:126]), horiz=T, names=rev(totalresults$term.name[75:126]), las=1, cex.names = 0.6);
par(mar=c(5,15,1,2));barplot(rev(-log10(totalresults$p.value[1:74])),xlab="-log10(p-value)",ylab="",cex.main=1.3, cex.lab=0.9, cex.axis=0.9, main="Over-Expressed Clusters", col=rev(totalcols[1:74]), horiz=T, names=rev(totalresults$term.name[1:74]), las=1, cex.names = 0.6);
```

