---
title: "R Practical -  Identifying Gene Clusters in the Linear Genome"
author: "Christoforos Nikolaou"
date: "`r Sys.Date()`"

output:
  html_document:
    code_folding: show
    fig_caption: yes
    toc: yes
    toc_depth: 3
    toc_float: yes
  word_document: null
  pdf_document:
    latex_engine: xelatex
link-citations: yes
bibliography: null
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
setwd("C:/Users/chris/Dropbox/github/MolBioMedClass/RMarkdown")
```


## Introduction

This is a first practical aiming at showcasing observed over expected frequencies

## test

## The problem

Description of the problem

## Loading Libraries and Data

Next we will have to load the necessary libraries and packages onto R. Remember that if a package is not already installed in your machine you will need to install it with $install.packages("nameOfpackage")$. Otherwise you can load it directly as we show below.

```{r}
library(tidyverse)
```

As you see the tidyverse() package is a package that loads a number of other packages and libraries that help us treat, analyze and visualize data.

### Loading data

We will next load the data from a file that is already 

```{r}
nrodata<-read.delim("https://raw.githubusercontent.com/christoforos-nikolaou/MolBioMedClass/master/Datasets/SacCer2_All_NRO_values.tsv", header=T, sep="\t")
str(nrodata)
```

```{r}
hist(nrodata$GROValue, breaks=100)
upratio<-length(which(nrodata$GROValue>=0))/length(nrodata$GROValue)
upratio
```


```{r}
library(rle)
chroms<-as.factor(nrodata$chrom)
runs<-rle(c("0"))
for(chromosome in levels(chroms)){
  i<-which(nrodata$chrom==chromosome)
  values<-nrodata$GROValue[i]
  values<-ifelse(values >= 0, 1, 0)
  runs<-c(runs,rle(values))
}
```

```{r}
posclusters<-runs$lengths[which(runs$values==1)]
negclusters<-runs$lengths[which(runs$values==0)]
tabp<-as.data.frame(table(posclusters))
tabn<-as.data.frame(table(negclusters))
tabp$Freq<-tabp$Freq/length(runs$lengths)
colnames(tabp)[1]<-"Cluster Size"
tabn$Freq<-tabn$Freq/length(runs$lengths)
colnames(tabn)[1]<-"Cluster Size"
tabp$type<-"positive"
tabn$type<-"negative"
tab<-rbind(tabp, tabn)
tab$`Cluster Size`<-as.numeric(as.character(tab$`Cluster Size`))

#
library(ggplot2)
ggplot(tab, aes(x = as.numeric(`Cluster Size`), y = Freq, color= type)) +
  geom_point(size = 5, alpha = 0.6) +
  scale_color_manual(values = c("positive" = "firebrick4", "negative" = "steelblue4")) +
  scale_x_continuous(breaks = sort(unique(tab$`Cluster Size`))) +
  labs(title = "Cluster Size Distribution", x = "Cluster Size", y = "Frequency") +
  theme_minimal()
```

```{r}
p<-vector(mode="numeric", length=30)
for(k in 1:30){
  p[k]=0.5**k
}
tabr<-data.frame("Cluster Size"=1:30, "Freq"=p, "type"="random")
colnames(tabr)[1]<-"Cluster Size"
#
tab<-rbind(tab, tabr)
#
ggplot(tab, aes(x = as.numeric(`Cluster Size`), y = log10(Freq), color= type)) +
  geom_point(size = 5, alpha = 0.6) +
  scale_color_manual(values = c("positive" = "firebrick4", "negative" = "steelblue4", "random"="grey30")) +
  scale_x_continuous(breaks = sort(unique(tab$`Cluster Size`))) +
  labs(title = "Cluster Size Distribution", x = "Cluster Size", y = "Frequency") +
  theme_minimal()
```

```{r}
clusterTable<-matrix(0, nrow=15, ncol=2)
for(i in 1:15){
  clusterTable[i,1]<-i
  clusterTable[i,2]<-tabp$Freq[i]/tabr$Freq[i]
}
clusterTable<-as.data.frame(clusterTable)
colnames(clusterTable)<-c("Cluster Size", "Probability Ratio")
ggplot(clusterTable, aes(x = as.numeric(`Cluster Size`), y = `Probability Ratio`)) +
  geom_point(size = 5, alpha = 0.6, color="olivedrab") +
  labs(title = "Cluster Size Enrichment", x = "Cluster Size", y = "Probability Ratio") +
  geom_hline(yintercept=1, color="grey50", linetype="dashed", linewidth=0.5) +
  theme_minimal()
```

```{r}
j<-which(runs$lengths>=7)
runs[j]
```

