{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "719d4a0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "In this exercise, we will deal with the identification of sequence patterns. Initially, we will try to identify small cis-elements in pre-aligned sequences, and then we will search for them in unknown sequences. \n",
    "The example of known sequences will be a set of binding sites for a transcription factor involved in the differentiation of hematopoietic system cells, GATA, as identified in the human genome. The data consist of 50 hexanucleotides, which you can download from the gata.fa file at the end of this page. \n",
    "(The data are taken from the book by Deonier, Tavare, and Waterman \"Computational Genome Analysis\").\n",
    "Exercise Objectives\n",
    "The objectives are:\n",
    "To read the aligned sequences into a data structure that we can manipulate\n",
    "To extract a consensus for the entire set\n",
    "To extract a PWM for the sequences at each position\n",
    "To graphically represent the PWM in two different ways.In this exercise, we will deal with the identification of sequence patterns. Initially, we will try to identify small cis-elements in pre-aligned sequences, and then we will search for them in unknown sequences. \n",
    "The example of known sequences will be a set of binding sites for a transcription factor involved in the differentiation of hematopoietic system cells, GATA, as identified in the human genome. The data consist of 50 hexanucleotides, which you can download from the gata.fa file at the end of this page. \n",
    "(The data are taken from the book by Deonier, Tavare, and Waterman \"Computational Genome Analysis\").\n",
    "Exercise Objectives\n",
    "The objectives are:\n",
    "To read the aligned sequences into a data structure that we can manipulate\n",
    "To extract a consensus for the entire set\n",
    "To extract a PWM for the sequences at each position\n",
    "To graphically represent the PWM in two different ways.\n",
    "1. Reading Sequences\n",
    "For the purpose of this exercise, we will use R, in which we will load the seqinr library (after setting the correct working folder).\n",
    "The steps are:\n",
    "Start R\n",
    "Declare the folder where you will download the data as the working directory with the command:\n",
    "setwd(\"...\");\n",
    "Load the function [readfastafile.R](https://www.dropbox.com/s/2hmtvg6wn9xgjst/readfastafile.R) as well as [seqMotif.R](https://www.dropbox.com/s/67815vjkrvxa72i/seqMotif.R) which you will find at the respective links\n",
    "source(\"seqMotif.R\")\n",
    "source(\"readfastafile.R\")\n",
    "At this point, we will read the GATA sequences file using the readfastafile function and store them in a sequence array called gata. You can download the file from [here](https://www.dropbox.com/s/pr51bnmp39b996a/gata.fa).\n",
    "gata<-readfastafile(\"gata.fa\")\n",
    "Examine the result to see the contents of the collection. It consists of 50 six-nucleotide sequences each. The fact that all sequences have the same length will be useful later for extracting pattern characteristics, as we will assume we can align the sequences one under the other and compare their residues by position.\n",
    "2. Extracting Consensus\n",
    "We now have the sequences stored in an R data structure. Our next goal is to extract a consensus sequence from the set of sequences in our collection. Remember that the consensus sequence corresponds to the most common residue extracted by position. You can extract the consensus sequence using the namesake function [consensus.R](https://www.dropbox.com/s/0n24xp9gjen6p1g/consensus.R) as shown below:\n",
    "source(\"consensus.R\")\n",
    "consensus(gata, threshold=0.3)\n",
    "You get a sequence corresponding to the consensus. As discussed in class, the probability with which the most prevalent residue appears can vary greatly for each position, affecting the reliability of the consensus. In the above example, we set the threshold to 0.3, meaning that if the maximum value was greater than 0.3, it returns that specific residue. What would the results be if we wanted to be stricter, e.g., requiring the most prevalent residue to have a probability of at least 0.6?\n",
    "consensus(gata, threshold=0.6)\n",
    "You see how the consensus changes because for the last position there is relative ambiguity (no residue has p>=0.6). In this case, the function returns all residues with a probability of at least 1-threshold (in this case [AG]). Try increasing the threshold further and see how the consensus becomes increasingly \"probabilistic\". This does not mean you can no longer say what happens at each position, just that the initial estimate for a clear consensus is not so reliable.\n",
    "The best way to estimate a pattern is to look at the Probability Weight Matrix (PWM) and represent it as a logo\n",
    "3. PWM\n",
    "Beyond a simple consensus sequence, a sequence pattern can be quantified and analyzed better with a PWM, i.e., a position-specific probability matrix. Using the seqMotif.R function, we can calculate the PWM of a sequence collection as follows:\n",
    "seqMotif(gata, drawmotif=F)\n",
    "The result is the probability matrix where the preference for the \"GATA\" pattern in the four middle positions of the sequence is clear, but we see that the trend is not equivalent for all residues. For example, the fifth position of the sequence is A with probability 0.72, while the third (also A) has a corresponding probability of 0.96. \n",
    "In this case, we chose not to represent the motif with the option (drawmotif=F), but if we want we can see a first visualization of it in the form of a heatmap with the following command:\n",
    "seqMotif(gata, drawmotif=T)\n",
    "Observe the shape that emerges. You first see on the x-axis the nucleotide positions we are examining (from P1 to P6) and on the y-axis the four possible outcomes (nucleotide). The corresponding elements (base by position) are colored based on the PWM, with red values for high and white for low. Can you see the correspondence with the consensus?\n",
    "You can easily save the image you just created in a format readable by any computer (pdf) with the command:\n",
    "pdf(\"myfirstheatmap.pdf\");seqMotif(gata, drawmotif=T);dev.off()\n",
    "4. Graphical Representation of PWM. Sequence Logo\n",
    "An even better way to represent a pattern is the one derived from aligned sequences through Shannon information analysis and its representation as a sequence logo. This can be done very easily using online services, the most well-known of which is the Weblogo from the University of Berkeley: [http://weblogo.berkeley.edu/logo.cgi](http://weblogo.berkeley.edu/logo.cgi).\n",
    "To use it, we simply need to write all the aligned sequences into a plain text txt file and then paste them onto the Weblogo Server page.\n",
    "We write the gata sequences into a file (which we name temp.txt, but you can name it whatever you want) with the write.table command:\n",
    "write.table(as.vector(unlist(gata)), file=\"temp.txt\", row.names=F, quote=F, col.names=F)\n",
    "Outside the R environment, go to the folder we are working in, find the temp.txt file, open it with a simple text editor (e.g., notepad or equivalent), and remove the first line. \n",
    "Open a browser and go to the page [http://weblogo.berkeley.edu/logo.cgi](http://weblogo.berkeley.edu/logo.cgi), where you simply paste the sequences from temp.txt and select create logo.\n",
    "Observe the logo that was created. To what extent does it contain information you got from the PWM via the heatmap?\n",
    "Comparison with Random Sequences\n",
    "To better understand the strength of a PWM, i.e., how characteristic the pattern it contains is, it is good to compare it with a set of random sequences. You can do all the above steps on a set of random sequences of the same length (6 nucleotides) which you will find in the [random.fa](https://www.dropbox.com/s/102g7y0ortf43dq/random.fa) file. Examine the heatmap and logo of this collection and see the differences between them and the GATA elements\n",
    "Questions:\n",
    "1. Download the random.fa sequence set which you will find at the end of the page and perform all the analyses described above to get the corresponding PWM. The random.fa sequences are a collection of random sequences \n",
    "2. Create a PSSM matrix by comparing the gata.fa and random.fa collections. You need to calculate the binary logarithm of the ratio of the two motifs with gata in the numerator. A necessary addition is to add pseudocounts to both motifs to avoid indeterminacies like N/0 or log(0). An indicative command adds a small amount (0.01) to both parts of the ratio:\n",
    "pssm<-log2((gata_motif+0.01)/(random_motif+0.01))\n",
    "where gata_motif and random_motif have been obtained by applying seqMotif() to the respective sequence collections.\n",
    "3. Having created the PSSM, you need to search for this specific motif in the sequence given to you as [test1.fa](https://www.dropbox.com/s/ni6nrp0niv47kd2/test1.fa) at the end of the page. The steps to follow are:\n",
    "a. Read the sequence\n",
    "targetseq<-readfastafile(\"test1.fa\")\n",
    "b. download the [pssmSearch.R](https://www.dropbox.com/s/4ssz9kg3dj9bhbt/pssmSearch.R) function and execute it as follows:\n",
    "pssmSearch(pssm, targetseq, threshold=X)\n",
    "1. Reading Sequences\n",
    "For the purpose of this exercise, we will use R, in which we will load the seqinr library (after setting the correct working folder).\n",
    "The steps are:\n",
    "Start R\n",
    "Declare the folder where you will download the data as the working directory with the command:\n",
    "setwd(\"...\");\n",
    "Load the function [readfastafile.R](https://www.dropbox.com/s/2hmtvg6wn9xgjst/readfastafile.R) as well as [seqMotif.R](https://www.dropbox.com/s/67815vjkrvxa72i/seqMotif.R) which you will find at the respective links\n",
    "source(\"seqMotif.R\")\n",
    "source(\"readfastafile.R\")\n",
    "At this point, we will read the GATA sequences file using the readfastafile function and store them in a sequence array called gata. You can download the file from [here](https://www.dropbox.com/s/pr51bnmp39b996a/gata.fa).\n",
    "gata<-readfastafile(\"gata.fa\")\n",
    "Examine the result to see the contents of the collection. It consists of 50 six-nucleotide sequences each. The fact that all sequences have the same length will be useful later for extracting pattern characteristics, as we will assume we can align the sequences one under the other and compare their residues by position.\n",
    "2. Extracting Consensus\n",
    "We now have the sequences stored in an R data structure. Our next goal is to extract a consensus sequence from the set of sequences in our collection. Remember that the consensus sequence corresponds to the most common residue extracted by position. You can extract the consensus sequence using the namesake function [consensus.R](https://www.dropbox.com/s/0n24xp9gjen6p1g/consensus.R) as shown below:\n",
    "source(\"consensus.R\")\n",
    "consensus(gata, threshold=0.3)\n",
    "You get a sequence corresponding to the consensus. As discussed in class, the probability with which the most prevalent residue appears can vary greatly for each position, affecting the reliability of the consensus. In the above example, we set the threshold to 0.3, meaning that if the maximum value was greater than 0.3, it returns that specific residue. What would the results be if we wanted to be stricter, e.g., requiring the most prevalent residue to have a probability of at least 0.6?\n",
    "consensus(gata, threshold=0.6)\n",
    "You see how the consensus changes because for the last position there is relative ambiguity (no residue has p>=0.6). In this case, the function returns all residues with a probability of at least 1-threshold (in this case [AG]). Try increasing the threshold further and see how the consensus becomes increasingly \"probabilistic\". This does not mean you can no longer say what happens at each position, just that the initial estimate for a clear consensus is not so reliable.\n",
    "The best way to estimate a pattern is to look at the Probability Weight Matrix (PWM) and represent it as a logo\n",
    "3. PWM\n",
    "Beyond a simple consensus sequence, a sequence pattern can be quantified and analyzed better with a PWM, i.e., a position-specific probability matrix. Using the seqMotif.R function, we can calculate the PWM of a sequence collection as follows:\n",
    "seqMotif(gata, drawmotif=F)\n",
    "The result is the probability matrix where the preference for the \"GATA\" pattern in the four middle positions of the sequence is clear, but we see that the trend is not equivalent for all residues. For example, the fifth position of the sequence is A with probability 0.72, while the third (also A) has a corresponding probability of 0.96. \n",
    "In this case, we chose not to represent the motif with the option (drawmotif=F), but if we want we can see a first visualization of it in the form of a heatmap with the following command:\n",
    "seqMotif(gata, drawmotif=T)\n",
    "Observe the shape that emerges. You first see on the x-axis the nucleotide positions we are examining (from P1 to P6) and on the y-axis the four possible outcomes (nucleotide). The corresponding elements (base by position) are colored based on the PWM, with red values for high and white for low. Can you see the correspondence with the consensus?\n",
    "You can easily save the image you just created in a format readable by any computer (pdf) with the command:\n",
    "pdf(\"myfirstheatmap.pdf\");seqMotif(gata, drawmotif=T);dev.off()\n",
    "4. Graphical Representation of PWM. Sequence Logo\n",
    "An even better way to represent a pattern is the one derived from aligned sequences through Shannon information analysis and its representation as a sequence logo. This can be done very easily using online services, the most well-known of which is the Weblogo from the University of Berkeley: [http://weblogo.berkeley.edu/logo.cgi](http://weblogo.berkeley.edu/logo.cgi).\n",
    "To use it, we simply need to write all the aligned sequences into a plain text txt file and then paste them onto the Weblogo Server page.\n",
    "We write the gata sequences into a file (which we name temp.txt, but you can name it whatever you want) with the write.table command:\n",
    "write.table(as.vector(unlist(gata)), file=\"temp.txt\", row.names=F, quote=F, col.names=F)\n",
    "Outside the R environment, go to the folder we are working in, find the temp.txt file, open it with a simple text editor (e.g., notepad or equivalent), and remove the first line. \n",
    "Open a browser and go to the page [http://weblogo.berkeley.edu/logo.cgi](http://weblogo.berkeley.edu/logo.cgi), where you simply paste the sequences from temp.txt and select create logo.\n",
    "Observe the logo that was created. To what extent does it contain information you got from the PWM via the heatmap?\n",
    "Comparison with Random Sequences\n",
    "To better understand the strength of a PWM, i.e., how characteristic the pattern it contains is, it is good to compare it with a set of random sequences. You can do all the above steps on a set of random sequences of the same length (6 nucleotides) which you will find in the [random.fa](https://www.dropbox.com/s/102g7y0ortf43dq/random.fa) file. Examine the heatmap and logo of this collection and see the differences between them and the GATA elements\n",
    "Questions:\n",
    "1. Download the random.fa sequence set which you will find at the end of the page and perform all the analyses described above to get the corresponding PWM. The random.fa sequences are a collection of random sequences \n",
    "2. Create a PSSM matrix by comparing the gata.fa and random.fa collections. You need to calculate the binary logarithm of the ratio of the two motifs with gata in the numerator. A necessary addition is to add pseudocounts to both motifs to avoid indeterminacies like N/0 or log(0). An indicative command adds a small amount (0.01) to both parts of the ratio:\n",
    "pssm<-log2((gata_motif+0.01)/(random_motif+0.01))\n",
    "where gata_motif and random_motif have been obtained by applying seqMotif() to the respective sequence collections.\n",
    "3. Having created the PSSM, you need to search for this specific motif in the sequence given to you as [test1.fa](https://www.dropbox.com/s/ni6nrp0niv47kd2/test1.fa) at the end of the page. The steps to follow are:\n",
    "a. Read the sequence\n",
    "targetseq<-readfastafile(\"test1.fa\")\n",
    "b. download the [pssmSearch.R](https://www.dropbox.com/s/4ssz9kg3dj9bhbt/pssmSearch.R) function and execute it as follows:\n",
    "pssmSearch(pssm, targetseq, threshold=X)"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
